pkg/project/controller/controller.go

```diff
@@ -23,17 +25,61 @@ type fatalError string
 func (e fatalError) Error() string { return "fatal error handling namespace: " + string(e) }
 
 // Handle processes a namespace and deletes content in origin if its terminating
-func (c *NamespaceController) Handle(namespace *kapi.Namespace) error {
-	// ignore namespaces that are not terminating
+func (c *NamespaceController) Handle(namespace *kapi.Namespace) (err error) {
+	// if namespace is not terminating, ignore it
 	if namespace.Status.Phase != kapi.NamespaceTerminating {
 		return nil
 	}
 
-	deleteAllContent(c.Client, namespace.Name)
-	// TODO: finalize namespace (remove openshift.com/origin)
+	// if we already processed this namespace, ignore it
+	if finalized(namespace) {
+		return nil
+	}
+
+	// there may still be content for us to remove
+	err = deleteAllContent(c.Client, namespace.Name)
+	if err != nil {
+		return err
+	}
+
+	// we have removed content, so mark it finalized by us
+	err = finalize(c.KubeClient, namespace)
+	if err != nil {
+		return err
+	}
+
 	return nil
 }
 
+// finalized returns true if the spec.finalizers does not contain the project finalizer
+func finalized(namespace *kapi.Namespace) bool {
+	for i := range namespace.Spec.Finalizers {
+		if api.FinalizerProject == namespace.Spec.Finalizers[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// finalize will finalize the namespace for kubernetes
+func finalize(kubeClient kclient.Interface, namespace *kapi.Namespace) error {
+	namespaceFinalize := kapi.Namespace{}
+	namespaceFinalize.ObjectMeta = namespace.ObjectMeta
+	namespaceFinalize.Spec = namespace.Spec
+	finalizerSet := util.NewStringSet()
+	for i := range namespace.Spec.Finalizers {
+		if namespace.Spec.Finalizers[i] != api.FinalizerProject {
+			finalizerSet.Insert(string(namespace.Spec.Finalizers[i]))
+		}
+	}
+	namespaceFinalize.Spec.Finalizers = make([]kapi.FinalizerName, 0, len(finalizerSet))
+	for _, value := range finalizerSet.List() {
```

Comment:
Order does not matter.

I should change upstream to make sure the field cannot have duplicate values though.

Will make that change in Kube repo

Sent from my iPhone

> On Apr 6, 2015, at 3:53 PM, David Eads <notifications@github.com> wrote:
> 
> In pkg/project/controller/controller.go:
> 
> > +	return true
> > +}
> > +
> > +// finalize will finalize the namespace for kubernetes
> > +func finalize(kubeClient kclient.Interface, namespace *kapi.Namespace) error {
> > +	namespaceFinalize := kapi.Namespace{}
> > +	namespaceFinalize.ObjectMeta = namespace.ObjectMeta
> > +	namespaceFinalize.Spec = namespace.Spec
> > +	finalizerSet := util.NewStringSet()
> > +	for i := range namespace.Spec.Finalizers {
> > +		if namespace.Spec.Finalizers[i] != api.FinalizerProject {
> > +			finalizerSet.Insert(string(namespace.Spec.Finalizers[i]))
> > +		}
> > +	}
> > +	namespaceFinalize.Spec.Finalizers = make([]kapi.FinalizerName, 0, len(finalizerSet))
> > +	for _, value := range finalizerSet.List() {
> Does the order of finalizers matter? This effectively reorders the list.
> 
> —
> Reply to this email directly or view it on GitHub.
>